### PR TITLE
cpu/stm32f0f1f3: add MCO configuration and initialization

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR extends the STM32F0/F1/F3 clock config/initialization with support for MCO (microcontroller clock output). This feature allows to output one clock source among HSE, HSI, PLLCLK / 2 or SYSCLK to a GPIO pin (PA8 with AF0).

On F0 and F3, MCO prescaler can be configured (among 1, 2, 4, 8, 16, the default is 1). No prescaler is available on F1.

LSE/LSE input clocks are not configurable although they could on F0 and F3. This can be done in a follow-up PR.

Using one of the MCO clock source will automatically enable the corresponding underlying clock (HSE, HSI or PLL). So for example, it's possible to use HSI as system clock while using the PLLCLK / 2 as input clock for MCO.

### Testing procedure

To test this PR, one needs a logic analyzer or a scope to verify the clock output on PA8 when MCO is enabled. I'm using a chinese saelae logic analyzer with PulseView on Ubuntu. In this setup, I cannot have a clean measure for clock frequency > 24MHz because my logic analyzer has a maximum sample rate of 48MHz.

Here I provide the output of MCO measured with a nucleo-f072rb, nucleo-f103rb and nucleo-f303re, PA8 is available on D7. Any application can be used, so here I'm testing with examples/hello-world.

- **nucleo-f072rb**:

<details><summary>MCO: HSI => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSI=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f072rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSE=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f072rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: PLLCLK / 2 + MCO prescaler 8 + SYSCLK = HSE=> expected frequency is 3MHz (PLLCLK is 48MHz, but PLLCLK /2 is given so 24 / 8)</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_PLLCLK=1 -DCONFIG_CLOCK_MCO_PRE=8" BUILD_IN_DOCKER=1  make BOARD=nucleo-f072rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: SYSCLK + SYSCLK = HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_SYSCLK=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f072rb -C examples/hello-world flash term --no-print-directory
```

</details>

- **nucleo-f103rb**:

<details><summary>MCO: HSI => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSI=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f103rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f103rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: PLLCLK / 2 + PLLMUL=3 + SYSCLK = HSE=> expected frequency is 12MHz (PLLCLK = (8MHz / 1) * 3 = 24MHz and MCO receives PLLCLK / 2 => <strong>12MHz</strong>)</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_PLLCLK=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f103rb -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: SYSCLK + SYSCLK = HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_SYSCLK=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f103rb -C examples/hello-world flash term --no-print-directory
```

</details>

- **nucleo-f303re**:

<details><summary>MCO: HSI => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSI=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f303re -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_PLL=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_HSE=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f303re -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: PLLCLK / 2 + MCO prescaler 8 + SYSCLK = HSE=> expected frequency is 3MHz (PLLCLK is 48MHz, but PLLCLK /2 is given so 24 / 8)</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_PLLCLK=1 -DCONFIG_CLOCK_MCO_PRE=8" BUILD_IN_DOCKER=1  make BOARD=nucleo-f303re -C examples/hello-world flash term --no-print-directory
```

</details>

<details><summary>MCO: SYSCLK + SYSCLK = HSE => expected frequency is 8MHz</summary>

image

```
$ CFLAGS="-DCONFIG_USE_CLOCK_HSE=1 -DCONFIG_CLOCK_ENABLE_MCO=1 -DCONFIG_CLOCK_MCO_USE_SYSCLK=1 -DCONFIG_CLOCK_MCO_PRE=1" BUILD_IN_DOCKER=1  make BOARD=nucleo-f303re -C examples/hello-world flash term --no-print-directory
```

</details>

### Issues/PRs references

Tick one item in #14975
Contains a commit from #14946 and is based on #14923

### CODE REVIEW

**Review by fjmolinas — DISMISSED** [review]
ACK, the configuration matches the datasheet. I don't have the boards to test but I trust @aabadie detailed testing results.

File: cpu/stm32/stmclk/stmclk_f0f1f3.c [hunk 1]
```diff
@@ -95,11 +96,130 @@
...
+
+/* Configure the MCO prescaler: valid values are 1, 2, 4, 8, 16, 32, 64, 128 */
+#ifndef CONFIG_CLOCK_MCO_PRE
+#define CONFIG_CLOCK_MCO_PRE                    (1)
+#endif
+
+/* MCO prescaler is only available on STM32F0 and STM32F3 */
+#if defined(CPU_FAM_STM32F0) || defined(CPU_FAM_STM32F3)
+#if CONFIG_CLOCK_MCO_PRE == 1
+#define CLOCK_MCO_PRE                           (RCC_CFGR_MCOPRE_DIV1)
```
Comment (fjmolinas) [inline cpu/stm32/stmclk/stmclk_f0f1f3.c:178]: This one does not seem to exit for al `stm32f3`, e.g.: stm32f3discovery

Comment (aabadie) [reply cpu/stm32/stmclk/stmclk_f0f1f3.c:183]: In fact, the prescaler doesn't exists for some F3 lines. I adapted the code this way.

Comment (aabadie) [reply cpu/stm32/stmclk/stmclk_f0f1f3.c:183]: See RM0316, figure 13, page 126.

**Comment by fjmolinas** [comment]
@aabadie looks good please squash!

**Review by fjmolinas — APPROVED** [review]
ACK, trusting @aabadie testing output, and since the feature is disabled by default the current configuration is not affected.

**Comment by aabadie** [comment]
Hmm, looks like f0 also needs work :)

**Comment by aabadie** [comment]
I should now be fixed for f0 and f3: I added an explicit list of f0 and f3 lines with the prescaler (based on the presence of the `RCC_CFGR_MCOPRE_DIVXX` defines.

**Comment by fjmolinas** [comment]
Looks, good ACK still valid

**Pull request merged**